src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -225,9 +218,110 @@ private void pre20Upgrade() throws Exception {
+     * It will return only the states for indices that actually have a shard allocated on the current node.
+     *
+     * @param event           the cluster state event from which we figure out what is new in each index and should potentially be written
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStateOnDataNode(ClusterChangedEvent event, MetaData currentMetaData) {
+        ClusterState newState = event.state();
+        ClusterState previousState = event.previousState();
+
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        RoutingNode newRoutingNode = newState.getRoutingNodes().node(newState.nodes().localNodeId());
+        RoutingNode previousRoutingNode = previousState.getRoutingNodes().node(newState.nodes().localNodeId());
+        if (newRoutingNode == null) {
+            // this needs some other handling
+            return indicesToWrite.values();
+        }
+        // iterate over all shards allocated on this node in the new cluster state but only write if ...
+        for (MutableShardRouting shardRouting : newRoutingNode) {
+            IndexMetaData indexMetaData = newState.metaData().index(shardRouting.index());
+            IndexMetaData currentIndexMetaData = currentMetaData == null ? null : currentMetaData.index(indexMetaData.index());
+            String writeReason = null;
+            // ... state persistence was disabled or index was newly created
+            if (currentIndexMetaData == null) {
+                writeReason = "freshly created";
+                // ... new shard is allocated on node (we could optimize here and make sure only written once and not for each shard per index -> do later)
+            } else if (shardsOfThisIndexExistedOnNodeAlready(previousRoutingNode, shardRouting) == false) {
+                writeReason = "newly allocated on node";
+                // ... version changed
+            } else if (indexMetaData.version() != currentIndexMetaData.version()) {
+                writeReason = "version changed from [" + currentIndexMetaData.version() + "] to [" + indexMetaData.version() + "]";
+            }
+            if (writeReason != null) {
+                indicesToWrite.put(shardRouting.index(),
+                        new GatewayMetaState.IndexMetaWriteInfo(indexMetaData, currentIndexMetaData,
+                                writeReason));
+            }
+        }
+        return indicesToWrite.values();
+    }
+
+    public static boolean shardsOfThisIndexExistedOnNodeAlready(RoutingNode previousRoutingNode, MutableShardRouting shardRouting) {
+        return previousRoutingNode.shardsWithState(shardRouting.index(), ShardRoutingState.INITIALIZING, ShardRoutingState.STARTED).size() != 0;
```

Comment:
this is tricky as it misses `ShardRoutingState#RELOCATING`. Can we add a method that just returns all the shards of an index? or alternatively, if we follow  my previously suggestion of caching written indices - just check the new index if it is not part of the written indices set.